### PR TITLE
Refactoring TraceHeader Calls

### PR DESCRIPTION
**Pull request by atshaw43**
*Issue #, if available:*

*Description of changes:*
**Problem 1**
We have redundant code for creating a trace header.

**Solution**
Create a centralized function for creating trace headers from entities. This required NoOpSubsegment to have its traceId.

**Problem 2**
Why does NoOpSubsegment need a namespace and name? We have a bug where we are not correctly ending subsegments if they are NoOp. The subsegment gets created, but never ended by afterResponse because it is seen as a duplicate because its namespace and name is always empty.

**Solution**
Pass down the name and namespace in NoOpSubsegments.

https://github.com/aws/aws-xray-sdk-java/blob/master/aws-xray-recorder-sdk-aws-sdk/src/main/java/com/amazonaws/xray/handlers/TracingHandler.java

```
@Override
    public void afterResponse(Request<?> request, Response<?> response) {
        if (isSubsegmentDuplicate(recorder.getCurrentSubsegmentOptional(), request)) {
            Optional<Subsegment> currentSubsegmentOptional = recorder.getCurrentSubsegmentOptional();
            if (!currentSubsegmentOptional.isPresent()) {
                return;
            }
            Subsegment currentSubsegment = currentSubsegmentOptional.get();
            populateAndEndSubsegment(currentSubsegment, request, response, null);
        }
    }
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by atshaw43** [comment]
Problem: Why does NoOpSubsegment need a namespace? We have a bug where we are not correctly ending subsegments if they are NoOp. The subsegment gets created, but never ended by afterResponse because it is seen as a duplicate because its namespace is always empty.

Solution: Pass down the namespace in NoOpSubsegments.

https://github.com/aws/aws-xray-sdk-java/blob/master/aws-xray-recorder-sdk-aws-sdk/src/main/java/com/amazonaws/xray/handlers/TracingHandler.java

@Override
    public void afterResponse(Request<?> request, Response<?> response) {
        if (isSubsegmentDuplicate(recorder.getCurrentSubsegmentOptional(), request)) {
            Optional<Subsegment> currentSubsegmentOptional = recorder.getCurrentSubsegmentOptional();
            if (!currentSubsegmentOptional.isPresent()) {
                return;
            }
            Subsegment currentSubsegment = currentSubsegmentOptional.get();
            populateAndEndSubsegment(currentSubsegment, request, response, null);
        }
    }

File: aws-xray-recorder-sdk-aws-sdk/src/main/java/com/amazonaws/xray/handlers/TracingHandler.java [hunk 1]
```diff
@@ -194,15 +193,8 @@ public void beforeRequest(Request<?> request) {
...
-            boolean isSampled = previousSubsegment.isPresent() ?
-                    previousSubsegment.get().isSampled() :
-                    recorder.getCurrentSegment().isSampled();
-
-            TraceHeader header =
-                new TraceHeader(recorder.getCurrentSegment().getTraceId(),
-                        isSampled ? currentSubsegment.getId() : null,
-                        isSampled ? SampleDecision.SAMPLED : SampleDecision.NOT_SAMPLED);
-            request.addHeader(TraceHeader.HEADER_KEY, header.toString());
+            request.addHeader(TraceHeader.HEADER_KEY, TraceHeader.fromEntity(
```
Comment (srprash) [inline aws-xray-recorder-sdk-aws-sdk/src/main/java/com/amazonaws/xray/handlers/TracingHandler.java:196]: This code doesn't look equivalent to the one it's replacing.
In the left side code, the parent id as the TraceHeader argument is either `currentSubsegment`'s id or `null`. Only the `isSampled` is fetched from either previousSubsegment or the `recorder.getCurrentSegment()`.
In the right side code, the whole TraceHeader will be created either from `previousSubsegment` or `currentSubsegment`.

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/entities/NoOpSubSegment.java [hunk 2]
```diff
@@ -65,12 +79,14 @@ class NoOpSubSegment implements Subsegment {
...
 
-        this.isSampled = samplingStrategyOverride == SamplingStrategyOverride.DISABLED ?
-                parentSegment.isSampled() :
-                false;
-        this.isRecording = isSampled;
+        this.isSampled = false;
+        this.isRecording = false;
         this.samplingStrategyOverride = samplingStrategyOverride;
+
+        id = creator.getIdGenerator().newEntityId();
```
Comment (srprash) [inline aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/entities/NoOpSubSegment.java:89]: If this `NoOpSubsegment` is never emitted and never propagated, can we set a static id here? We should avoid unnecessary computation if we can.

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/entities/Segment.java [hunk 3]
```diff
@@ -43,6 +43,8 @@ static Segment noOp(TraceID traceId, AWSXRayRecorder recorder) {
     /**
      * @return the sampled
      */
+    @JsonIgnore
+    @Override
```
Comment (srprash) [inline aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/entities/Segment.java:47]: Do we need to necessarily override the `isSampled()` method from `Entity`?
Is it possible for `Entity` types and subtypes to directly use this method from `Entity` without an override?

File: aws-xray-recorder-sdk-aws-sdk/src/main/java/com/amazonaws/xray/handlers/TracingHandler.java (hunk 1 above)
Comment (atshaw43) [reply aws-xray-recorder-sdk-aws-sdk/src/main/java/com/amazonaws/xray/handlers/TracingHandler.java]: Wow, good catch! This was a bug.
I looked at the original code and I don't think we even need to use the previous subsegment here.

**Review by srprash — APPROVED** [review]

**Pull request merged**